pyteal/ast/frame.py

```diff
@@ -11,8 +14,101 @@
     from pyteal.compiler import CompileOptions
 
 
+@dataclass
+class LocalTypeSegment:
+    local_type: TealType
+    cnt: int
+    auto_instance: Expr = field(init=False)
+
+    def __post_init__(self):
+        if self.cnt <= 0:
+            raise TealInternalError(
+                "LocalTypeSegment initialization error: segment length must be strictly greatly than 0."
+            )
+        match self.local_type:
+            case TealType.uint64 | TealType.anytype:
+                self.auto_instance = Int(0)
+            case TealType.bytes:
+                self.auto_instance = Bytes("")
+            case TealType.none:
+                raise TealInternalError(
+                    "Local variable in subroutine initialization must be typed."
+                )
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        if self.cnt == 1:
+            inst_srt, inst_end = self.auto_instance.__teal__(options)
+            return inst_srt, inst_end
+        else:
+            dupn_srt, dupn_end = DupN(self.auto_instance, self.cnt).__teal__(options)
+            return dupn_srt, dupn_end
+
+
+LocalTypeSegment.__module__ = "pyteal"
+
+
+@dataclass
+class ProtoStackLayout:
+    arg_stack_types: list[TealType]
+    local_stack_types: list[TealType]
+    has_output: bool
+    output_index: Optional[int] = field(init=False)
+    succinct_repr: list[LocalTypeSegment] = field(init=False)
+
+    def __post_init__(self):
+        if self.has_output and len(self.local_stack_types) == 0:
+            raise TealInternalError(
+                "ProtoStackLayout initialization error: cannot output without local variable allocs."
+            )
+
+        self.output_index = 0 if self.has_output else None
+        if not all(map(lambda t: types_match(t, TealType.none), self.arg_stack_types)):
+            raise TealInternalError("Variables in frame memory layout must be typed.")
+
+        # Type check of local variables are performed over LocalTypeSegments
+        self.succinct_repr = [
+            LocalTypeSegment(t_type, len(list(dup_seg)))
+            for t_type, dup_seg in groupby(self.local_stack_types)
+        ]
+
+        # anytype at i + 1 merge to i
+        for i in reversed(range(len(self.succinct_repr) - 1)):
+            if self.succinct_repr[i + 1].local_type == TealType.anytype:
+                self.succinct_repr[i].cnt += self.succinct_repr[i + 1].cnt
+                self.succinct_repr.pop(i + 1)
```

Comment:
My gut is telling me that this will cause gaps and therefore be faulty in the case that there are consecutive non-`anytype`s in the middle of the sequence. However, a test case such as the following added to `SuccinctReprTestCase` could prove my gut wrong:

```python
        SuccinctReprTestCase(
            [
                pt.TealType.anytype,
                pt.TealType.bytes,
                pt.TealType.bytes,
                pt.TealType.uint64,
                pt.TealType.uint64,
                pt.TealType.uint64,
                pt.TealType.anytype,
                pt.TealType.anytype,
                pt.TealType.anytype,
                pt.TealType.anytype,
                pt.TealType.bytes,
                pt.TealType.bytes,
                pt.TealType.uint64,
                pt.TealType.uint64,
                pt.TealType.uint64,
            ],
            [
                LocalTypeSegment(pt.TealType.bytes, 3),
                LocalTypeSegment(pt.TealType.uint64, 7),
                LocalTypeSegment(pt.TealType.bytes, 2),
                LocalTypeSegment(pt.TealType.uint64, 3),
            ],
        ),
```

Reply:
good testcase, add to frame_test and seems to be working!

Reply:
Just to leave this thread open: I think I am tryin' to act smart here by merging `TealType.anytype` to nearest `TealType.byte` or `TealType.uint64`. But does that worth it? People feel free to persuade me on not worth it.

Reply:
I would benefit from an explanation of the motivation for the algorithm. This has to do with my "pros/cons" comment above. But if the pro is strong enough, and the tests give confidence that there are no bugs, then I would argue that it's "worth it".

Reply:
I'm somewhat confused by this sort of "casting" into a type based on adjacent types, why not keep them anytype?

Reply:
Uh I missed discussing the algorithm, my bad.

The motivation in applying the algorithm is to generate smallest code for local var init.

The expected IO of the algorithm follows:
- given a "sucinct" expression of local variable type description, e.g.,
  `[(TealType.uint64, 3), (TealType.anytype, 4), (TealType.bytes, 5)]`
- output a "shrinked" succinct expression of local variable type description, e.g., 
  `[(TealType.uint64, 7), (TealType.bytes, 5)]`

Now I am thinking out loud:
- How much it saved? Probably at most 2 ops for each occurrence of `TealType.anytype`.
- Is it somewhat confusing? I somehow see the confusion in maintaining the code.

Given the complexity (maybe unnecessary) and profit gain from smaller code, I am slightly prone to taking the algorithm out.

Reply:
I just take the algorithm out, at a point I am unconvinced that the algorithm brings more benefit.